packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -886,6 +893,120 @@ export class ProjectService {
+     * will be expanded to support this behavior natively.
+     */
+    private static async _compileMetricQueryWithNewTableCalculationsEngine(
+        metricQuery: MetricQuery,
+        explore: Explore,
+        warehouseClient: WarehouseClient,
+        userAttributes: UserAttributeValueMap,
+        granularity?: DateGranularity,
+    ): Promise<[CompiledQuery, CompiledQuery]> {
+        const exploreWithOverride = ProjectService.updateExploreWithGranularity(
+            explore,
+            metricQuery,
+            warehouseClient,
+            granularity,
+        );
+
+        const {
+            compiledMetricQueryWithoutTableCalculations,
+            compiledTableCalculations,
+            tableCalculationFilters,
+        } = ProjectService.isolateTableCalculationsFromCompiledMetricsQuery(
+            compileMetricQuery({
+                explore: exploreWithOverride,
+                metricQuery,
+                warehouseClient,
+            }),
+        );
+
+        /**
+         * Generate a new SELECT statement with all of our original columns, as well as
+         * table calculation columns/aggregates:
+         */
+        const selectFrom = [
+            '*',
+            ...compiledTableCalculations.map(
+                ({ name, compiledSql }) => `${compiledSql} AS ${name}`,
+            ),
+        ];
+
+        /**
+         * Render table calculation filter rules for compatibility with DuckDB.
+         */
+        const filterRules = getFilterRulesFromGroup(
+            tableCalculationFilters as FilterGroup,
+        );
+
+        const renderedFilters = filterRules.map((filterRule) => {
+            const field = compiledTableCalculations.find(
+                ({ name }) =>
+                    `table_calculation_${name}` === filterRule.target.fieldId,
+            );
+
+            /**
+             * If a matching field cannot be found, we insert a placeholder expression.
+             */
+            if (!field) {
+                return '1=1';
+            }
+
+            return renderTableCalculationFilterRuleSql(
+                filterRule,
+                field,
+                '"',
+                "'",
+                '\\',
+            );
+        });
+
+        const whereClause =
+            renderedFilters.length > 0
+                ? `WHERE ${renderedFilters.join('\nAND ')}`
+                : '';
+
+        const sorts = compiledMetricQueryWithoutTableCalculations.sorts.map(
+            ({ descending, fieldId }) =>
+                `${fieldId} ${descending ? 'DESC' : 'ASC'}`,
+        );
+
+        const orderByClause =
+            sorts.length > 0 ? `ORDER BY ${sorts.join(', ')}` : '';
+
+        const primaryQuery = buildQuery({
+            explore: exploreWithOverride,
+            compiledMetricQuery: compiledMetricQueryWithoutTableCalculations,
+            warehouseClient,
+            userAttributes,
+        });
+
+        const tableCalculationsSubQuery: CompiledQuery = {
```

Comment:
Most of this can probably be done via the existing query builder logic, with some refactoring.